### PR TITLE
Finish Sorting Assignment for Nathan Thomas 🦁

### PR DESCRIPTION
**Pull request by nwthomas**
*MVP:*
- [x] Finish implementing `selection_sort()` using the described specifications
- [x] Craft `bubble_sort()` via the listed requirements in the `README.md`
- [x] Ensure that Day 1 MVP items pass associated tests
- [x] Build out full `merge_sort()` functionality for merge, merge_sort, merge_in_place, merge_sort_in_place, and timsort
- [x] Use test file to make sure that all tests pass for `merge_sort()`

*STRETCH:*
- [ ] Build out `count_sort()` function
- [x] Update with full `linear_search()` function
- [x] Build non-recursive `binary_search()`
- [x] Implement recursive `binary_search()`
- [ ] Create `timsort()` algorithm
- [ ] Make an `O(1)` space complexity "in-place" `merge_sort()` algorithm

### CODE REVIEW

File: src/iterative_sorting/iterative_sorting.py [hunk 1]
```diff
@@ -1,30 +1,37 @@
...
-             
-
-
-
-        # TO-DO: swap
-
-
-
-
+        for j in range(i + 1, len(arr)):
```
Comment (andrewlowell) [inline src/iterative_sorting/iterative_sorting.py]: This is a perfectly serviceable solution and frankly, it's so concise you might have convinced me to prefer it :thinking: I believe the intended solution was to iterate through range(i + 1, len(arr)) like you did and keep track of the index of the smallest number in smallest_index, updating it as necessary. The nice thing is that in the worst case you would do O(n) swaps, one per index minus the last one. The solution you chose has a possible worst case O(n log n) swaps, is that true...? Worst case one swap for each iteration in the second for loop, so (n) + (n-1) + (n-2) + ... (1).

File: src/recursive_sorting/recursive_sorting.py [hunk 2]
```diff
@@ -1,17 +1,36 @@
-# TO-DO: complete the helpe function below to merge 2 sorted arrays
-def merge( arrA, arrB ):
-    elements = len( arrA ) + len( arrB )
-    merged_arr = [0] * elements
-    # TO-DO
-    
+from math import floor
+
+
+def merge(listA, listB):
```
Comment (andrewlowell) [inline src/recursive_sorting/recursive_sorting.py]: I also prefer the append to empty array solution so you can append the entire remaining array if one runs out.

File: src/recursive_sorting/recursive_sorting.py [hunk 3]
```diff
@@ -1,17 +1,36 @@
...
 
 
-# TO-DO: implement the Merge Sort function below USING RECURSION
-def merge_sort( arr ):
-    # TO-DO
+def merge_sort(list):
+    if len(list) <= 1:
+        return list
+    else:
+        pivot_index = floor(len(list) / 2)
```
Comment (andrewlowell) [inline src/recursive_sorting/recursive_sorting.py]: Any reason you prefer this to len(list) // 2?

Comment (nwthomas) [reply src/recursive_sorting/recursive_sorting.py]: No, I just didn't know floor division existed out of the box in Python when I wrote this. 😂

File: src/iterative_sorting/iterative_sorting.py (hunk 1 above)
Comment (nwthomas) [reply src/iterative_sorting/iterative_sorting.py]: Good point - I'll look into that. Do you want me to change it in a new PR?

**Pull request merged**